src/main/clojure/clara/rules/engine.cljc

```diff
@@ -1867,7 +1885,8 @@
   (fire-rules [session opts]
 
     (let [transient-memory (mem/to-transient memory)
-          transient-listener (l/to-transient listener)]
+          transient-listener (l/to-transient listener)
+          max-cycles (get opts :max-cycles 600000)]
```

Comment:
I'm not sure on the default chosen here. It could be breaking. 

I haven't thought it completely through, but is the counting no specifically counting cycles of rules happening, but instead just steps happening during fire-rules? Perhaps that is not granular enough.

Reply:
The count is intended to be incremented each time an activation group is fully fired.  I'm expecting that to be much smaller than the number of times a rule is fired, and so safer in terms of avoiding false positives and incorrectly throwing.

Reply:
I'll have to post more details later on how I chose the count and maybe change it - I measured the count for 30 seconds of a simple case on my device machine.  Open to discussion on what the count should be.

Reply:
I think I agree with Mike here:
> I'm not sure on the default chosen here. It could be breaking.

Given the fabricated set of rules
```
(ns clara.loop-detection
  (:require [clojure.test :refer :all]
            [clara.rules :as r]
            [clara.rules.accumulators :as acc]))

(defrecord Total [x])

(defrecord Subs [y])
(defrecord Sum [z])

(r/defrule sum-rule
  [Total (= ?total x)]
  [?s <- (acc/sum :y) :from [Subs]]
  [:test (> ?total ?s)]
  =>
  (r/insert! (->Sum (- ?total ?s))))

(r/defrule subs-rule
  [Sum]
  =>
  (r/insert-unconditional! (->Subs 1)))

(r/defquery sum-query
  []
  [?s <- (acc/count) :from [Subs]])

(-> (r/mk-session)
    (r/insert (->Total 600000))
    (r/fire-rules)
    (r/query sum-query))
```
This will break due to loop detection, when in fact it is a finite execution. While I won't expect to find this exact pattern in rules that we(cerner) have, I wouldn't be surprised if someone was doing something similar or using unconditional inserts to break loops.

On one hand, i would think that the addition of this would maintain the previous functionality of the rules engine by default(loop forever, or till done), but on the other I understand that someone just starting clara might not be aware of the functionality thus miss out on the detection option. 

Perhaps a large default is correct, but also allowing a sentinel value to disable loop detection to maintain passivity.

Reply:
My reasoning here was that it just didn't seem likely to hit it, but I see your points that it is always going to be possible and there is enough usage of Clara for odd use cases like this to exist.

How would we feel about something like the following?

- Having the limit be configurable, as is.  We could revisit the limit as well.
- Making the behaviour on reaching the limit configurable, with the options being to throw an exception, log a warning, or do nothing, with the log option being the default.  If a warning just went to standard out that would probably take care of most of the confusion on the part of users who misunderstand looping behaviour in Clara, while being less likely to cause problems for users than simply throwing an exception.  That said, the problems wouldn't go down to zero since if it bubbled up to the end-user (not a developer) in an app of some kind it could be confusing, as well as if Clara was in an app that had specific requirements around printing to standard out.
- In the future possibly adding a comprehensive "development mode" that would do things like enable throwing here, enable tracing, etc.  This would be an easier thing to explain in the docs than lots of different options requiring an understanding of Clara to make good choices.

If we go down the logging path, [tools.logging](https://github.com/clojure/tools.logging) might be an option.  I'm not delighted about adding dependencies, but it looks pretty lightweight and has no dependencies itself other than Clojure.  In essence it inspects the environment at runtime and uses whatever JVM logging framework is available, falling back to [java.util.logging])https://docs.oracle.com/javase/7/docs/api/java/util/logging/package-summary.html) if nothing else is available.  In ClojureScript I think we'd probably need to just print to the console, but I don't think the JS console is exposed to website users who aren't using developer tools so perhaps that isn't a major issue.

Sorry to take so long to respond here... I've been a bit preoccupied lately.

Thoughts @EthanEChristian @mrrodriguez ?

Reply:
I think that approach seems reasonable to me, particularly:
> Making the behavior on reaching the limit configurable

I feel like this allows us to have this functionality and still maintain passivity for consumers that have "odd" rules.

I don't think I have any objections to the addition of logging to clara's dependencies.

Reply:
@WilliamParker I like the default to be logging, however I don't think this should require a new dependency. 

I think it may actually be best to expose 2 args for this feature:

`max-cycles` - what you already have
`max-cycles-exceeded-fn` - tentative name, but it gives the consumer the ability to decide what to do

If they do not provide anything, perhaps we do a "print once" default fn. We can also provide a fn that can be used to throw the exception if that is what someone wants. These can be exposed as plain fn's or "special" keywords the argument accepts.

```clj
:max-cycles 60000
:max-cycles-exceeded-fn r/max-cycles-exceeded-print-once
;; Alternatives:
;; Special kw :clara-rules/max-cycles-exceeded-print-once
;; r/max-cycles-exceeded-throw
```
If no `:max-cycles-exceeded-fn` is given, use a default.

If we go this path, we should consider what args to this fn would be useful.

I think giving a handler to the consumer for cases like this is a common clj pattern.

@WilliamParker @EthanEChristian

Reply:
Per an offline conversation with @mrrodriguez, for my use-case the addition of the dependency wouldn't be an issue. After that conversation, I think that i can agree with Mike's hesitance on clara mandating a logger on it's consumers.

Reply:
Thanks for summarizing @EthanEChristian 

I'll re-iterate my concern is that I'm not a particular fan of adding dependencies in general in a general-use library - unless they are providing enough value to justify. Logging in a single case doesn't seem too compelling in that area.

Also, I know that `org.clojure/tools.logging` is quite flexible (I use it fairly often), but it can still lead to issues when added to any given classpath. It does do some implicit logger impl searching/loading and I think it could get particularly awkward if someone was using a logger framework that doesn't fit into `org.clojure/tools.logging`'s pattern at all. It may end up logging to it's own logger instead of their chosen one/may be annoying to try to silence or integrate with.

Reply:
Thanks for summarizing @EthanEChristian 

I'll re-iterate my concern is that I'm not a particular fan of adding dependencies in general in a general-use library - unless they are providing enough value to justify. Logging in a single case doesn't seem too compelling in that area.

Also, I know that `org.clojure/tools.logging` is quite flexible (I use it fairly often), but it can still lead to issues when added to any given classpath. It does do some implicit logger impl searching/loading and I think it could get particularly awkward if someone was using a logger framework that doesn't fit into `org.clojure/tools.logging`'s pattern at all. It may end up logging to it's own logger instead of their chosen one/may be annoying to try to silence or integrate with.

Reply:
OK it sounds like there is a consensus for a default behaviour of printing to standard out then.  I'll proceed with that.